File: backend/services/news_fetcher.py
"""
Financial news fetcher service.
Fetches relevant news articles from Google News RSS based on portfolio tickers,
account themes, and financial signals. No API key required.
"""
import logging
import urllib.request
import urllib.parse
import xml.etree.ElementTree as ET
from typing import List, Dict, Any, Optional
from datetime import datetime
import re
import hashlib

logger = logging.getLogger(__name__)

# Simple in-memory cache to avoid repeated fetches
_news_cache: Dict[str, tuple] = {}  # key -> (articles, fetched_at_timestamp)
_NEWS_CACHE_TTL = 1800  # 30 minutes


def _fetch_rss(query: str, max_results: int = 5) -> List[Dict[str, str]]:
    """Fetch articles from Google News RSS for a search query."""
    encoded_q = urllib.parse.quote(query)
    url = f"https://news.google.com/rss/search?q={encoded_q}&hl=en-US&gl=US&ceid=US:en"

    try:
        req = urllib.request.Request(url, headers={
            "User-Agent": "Mozilla/5.0 (compatible; NetworthPro/1.0)"
        })
        with urllib.request.urlopen(req, timeout=10) as response:
            xml_data = response.read()

        root = ET.fromstring(xml_data)
        articles = []

        for item in root.findall(".//item")[:max_results]:
            title_el = item.find("title")
            link_el = item.find("link")
            pub_date_el = item.find("pubDate")
            source_el = item.find("source")

            title = title_el.text if title_el is not None else ""
            link = link_el.text if link_el is not None else ""
            pub_date = pub_date_el.text if pub_date_el is not None else ""
            source = source_el.text if source_el is not None else ""

            if title and link:
                articles.append({
                    "title": _clean_title(title),
                    "url": link,
                    "source": source,
                    "published": _parse_rss_date(pub_date),
                })

        return articles

    except Exception as e:
        logger.warning(f"Failed to fetch news for '{query}': {e}")
        return []


def _clean_title(title: str) -> str:
    """Remove trailing source attribution from Google News titles."""
    # Google News titles often end with " - Source Name"
    parts = title.rsplit(" - ", 1)
    return parts[0].strip() if len(parts) == 2 and len(parts[1]) < 40 else title.strip()


def _parse_rss_date(date_str: str) -> str:
    """Parse RSS date into a readable format."""
    if not date_str:
        return ""
    try:
        # RSS dates look like: "Sat, 01 Feb 2026 14:30:00 GMT"
        dt = datetime.strptime(date_str.strip(), "%a, %d %b %Y %H:%M:%S %Z")
        return dt.strftime("%b %d, %Y")
    except (ValueError, TypeError):
        try:
            dt = datetime.strptime(date_str.strip()[:25], "%a, %d %b %Y %H:%M:%S")
            return dt.strftime("%b %d, %Y")
        except (ValueError, TypeError):
            return date_str[:20] if date_str else ""


def _build_search_queries(
    tickers: List[str],
    property_types: List[str],
    liability_categories: List[str],
    account_types: List[str],
) -> List[Dict[str, str]]:
    """Build targeted search queries from the user's financial data."""
    queries = []

    # Stock/ETF ticker queries -- most specific and relevant
    for ticker in tickers[:5]:  # Top 5 holdings
        ticker_clean = ticker.upper().strip()
        queries.append({
            "query": f"{ticker_clean} stock news",
            "theme": f"{ticker_clean} Stock",
        })

    # Sector-level queries based on ticker patterns
    sector_queries = set()
    tech_tickers = {"AAPL", "MSFT", "GOOGL", "GOOG", "META", "AMZN", "NVDA", "TSLA", "AMD", "INTC", "CRM", "NFLX"}
    finance_tickers = {"JPM", "BAC", "GS", "MS", "WFC", "V", "MA", "AXP", "BRK.B", "BRK.A"}
    health_tickers = {"JNJ", "UNH", "PFE", "MRK", "ABT", "TMO", "ABBV", "LLY"}
    energy_tickers = {"XOM", "CVX", "COP", "EOG", "SLB", "PSX"}

    ticker_set = set(t.upper() for t in tickers)
    if ticker_set & tech_tickers:
        sector_queries.add("technology sector market outlook")
    if ticker_set & finance_tickers:
        sector_queries.add("financial sector banking news")
    if ticker_set & health_tickers:
        sector_queries.add("healthcare sector pharma news")
    if ticker_set & energy_tickers:
        sector_queries.add("energy sector oil market news")

    # If user has broad ETFs, add market news
    etf_tickers = {"SPY", "VOO", "VTI", "QQQ", "IWM", "DIA", "VEA", "VWO", "BND", "AGG"}
    if ticker_set & etf_tickers:
        sector_queries.add("stock market outlook economy")

    for sq in list(sector_queries)[:2]:
        queries.append({"query": sq, "theme": "Market Outlook"})

    # Real estate themed queries
    if property_types:
        queries.append({
            "query": "real estate market housing prices",
            "theme": "Real Estate",
        })
        if "rental" in [pt.lower() for pt in property_types]:
            queries.append({
                "query": "rental property investment landlord news",
                "theme": "Rental Market",
            })

    # Debt/liability themed queries
    if liability_categories:
        cats_lower = [c.lower() for c in liability_categories if c]
        if any("student" in c for c in cats_lower):
            queries.append({"query": "student loan news policy", "theme": "Student Loans"})
        if any("credit" in c for c in cats_lower):
            queries.append({"query": "credit card interest rates personal finance", "theme": "Credit Cards"})
        if any("auto" in c for c in cats_lower):
            queries.append({"query": "auto loan rates car financing", "theme": "Auto Loans"})

    # General personal finance as a fallback
    if not queries:
        queries.append({"query": "personal finance investing tips", "theme": "Personal Finance"})

    return queries[:8]  # Cap total queries


def fetch_relevant_news(
    tickers: Optional[List[str]] = None,
    property_types: Optional[List[str]] = None,
    liability_categories: Optional[List[str]] = None,
    account_types: Optional[List[str]] = None,
    max_articles: int = 8,
) -> List[Dict[str, Any]]:
    """
    Fetch relevant financial news based on the user's portfolio and accounts.

    Returns:
        List of article dicts with 'title', 'url', 'source', 'published', 'theme'
    """
    # Build a cache key from inputs
    cache_input = f"{tickers}|{property_types}|{liability_categories}|{account_types}"
    cache_key = hashlib.md5(cache_input.encode()).hexdigest()

    # Check cache
    now = datetime.utcnow().timestamp()
    if cache_key in _news_cache:
        cached_articles, cached_at = _news_cache[cache_key]
        if now - cached_at < _NEWS_CACHE_TTL:
            logger.debug("Returning cached news articles")
            return cached_articles

    queries = _build_search_queries(
        tickers=tickers or [],
        property_types=property_types or [],
        liability_categories=liability_categories or [],
        account_types=account_types or [],
    )

    all_articles = []
    seen_titles = set()

    for q_info in queries:
        articles = _fetch_rss(q_info["query"], max_results=3)
        for article in articles:
            # Deduplicate by title similarity
            title_key = re.sub(r'[^a-z0-9]', '', article["title"].lower())[:50]
            if title_key not in seen_titles:
                seen_titles.add(title_key)
                article["theme"] = q_info["theme"]
                all_articles.append(article)

    result = all_articles[:max_articles]

    # Cache the result
    _news_cache[cache_key] = (result, now)

    # Prune old cache entries
    if len(_news_cache) > 50:
        expired = [k for k, (_, ts) in _news_cache.items() if now - ts > _NEWS_CACHE_TTL]
        for k in expired:
            del _news_cache[k]

    return result

File: frontend/lib/services/news-fetcher.ts
/**
 * Financial news fetcher service.
 * Fetches relevant news articles from Google News RSS based on portfolio tickers,
 * account themes, and financial signals. No API key required.
 */
import crypto from 'crypto';

export interface NewsArticle {
  title: string;
  url: string;
  source: string;
  published: string;
  theme: string;
}

// Simple in-memory cache
const _newsCache = new Map<string, { articles: NewsArticle[]; fetchedAt: number }>();
const NEWS_CACHE_TTL = 1800 * 1000; // 30 minutes in ms

function cleanTitle(title: string): string {
  const parts = title.split(' - ');
  if (parts.length === 2 && parts[1].length < 40) {
    return parts[0].trim();
  }
  return title.trim();
}

function parseRssDate(dateStr: string): string {
  if (!dateStr) return '';
  try {
    const dt = new Date(dateStr.trim());
    if (isNaN(dt.getTime())) return dateStr.slice(0, 20);
    return dt.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
  } catch {
    return dateStr.slice(0, 20);
  }
}

async function fetchRss(query: string, maxResults: number = 5): Promise<NewsArticle[]> {
  const encodedQ = encodeURIComponent(query);
  const url = `https://news.google.com/rss/search?q=${encodedQ}&hl=en-US&gl=US&ceid=US:en`;

  try {
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; NetworthPro/1.0)',
      },
      signal: AbortSignal.timeout(10000),
    });

    if (!response.ok) return [];

    const xmlData = await response.text();
    const articles: NewsArticle[] = [];

    // Simple XML parsing for RSS items
    const itemRegex = /<item>([\s\S]*?)<\/item>/g;
    let match;
    let count = 0;

    while ((match = itemRegex.exec(xmlData)) !== null && count < maxResults) {
      const itemXml = match[1];

      const titleMatch = /<title>([\s\S]*?)<\/title>/.exec(itemXml);
      const linkMatch = /<link>([\s\S]*?)<\/link>/.exec(itemXml);
      const pubDateMatch = /<pubDate>([\s\S]*?)<\/pubDate>/.exec(itemXml);
      const sourceMatch = /<source[^>]*>([\s\S]*?)<\/source>/.exec(itemXml);

      const title = titleMatch?.[1]?.replace(/<!\[CDATA\[(.*?)\]\]>/g, '$1').trim() || '';
      const link = linkMatch?.[1]?.trim() || '';
      const pubDate = pubDateMatch?.[1]?.trim() || '';
      const source = sourceMatch?.[1]?.replace(/<!\[CDATA\[(.*?)\]\]>/g, '$1').trim() || '';

      if (title && link) {
        articles.push({
          title: cleanTitle(title),
          url: link,
          source,
          published: parseRssDate(pubDate),
          theme: '', // Set by caller
        });
        count++;
      }
    }

    return articles;
  } catch (e) {
    console.warn(`Failed to fetch news for '${query}':`, e);
    return [];
  }
}

function buildSearchQueries(
  tickers: string[],
  propertyTypes: string[],
  liabilityCategories: string[],
  accountTypes: string[],
): Array<{ query: string; theme: string }> {
  const queries: Array<{ query: string; theme: string }> = [];

  // Stock/ETF ticker queries
  for (const ticker of tickers.slice(0, 5)) {
    const clean = ticker.toUpperCase().trim();
    queries.push({ query: `${clean} stock news`, theme: `${clean} Stock` });
  }

  // Sector-level queries
  const sectorQueries = new Set<string>();
  const techTickers = new Set(['AAPL', 'MSFT', 'GOOGL', 'GOOG', 'META', 'AMZN', 'NVDA', 'TSLA', 'AMD', 'INTC', 'CRM', 'NFLX']);
  const financeTickers = new Set(['JPM', 'BAC', 'GS', 'MS', 'WFC', 'V', 'MA', 'AXP']);
  const healthTickers = new Set(['JNJ', 'UNH', 'PFE', 'MRK', 'ABT', 'TMO', 'ABBV', 'LLY']);
  const energyTickers = new Set(['XOM', 'CVX', 'COP', 'EOG', 'SLB', 'PSX']);
  const etfTickers = new Set(['SPY', 'VOO', 'VTI', 'QQQ', 'IWM', 'DIA', 'VEA', 'VWO', 'BND', 'AGG']);

  const tickerSet = new Set(tickers.map((t) => t.toUpperCase()));

  if ([...tickerSet].some((t) => techTickers.has(t))) sectorQueries.add('technology sector market outlook');
  if ([...tickerSet].some((t) => financeTickers.has(t))) sectorQueries.add('financial sector banking news');
  if ([...tickerSet].some((t) => healthTickers.has(t))) sectorQueries.add('healthcare sector pharma news');
  if ([...tickerSet].some((t) => energyTickers.has(t))) sectorQueries.add('energy sector oil market news');
  if ([...tickerSet].some((t) => etfTickers.has(t))) sectorQueries.add('stock market outlook economy');

  for (const sq of [...sectorQueries].slice(0, 2)) {
    queries.push({ query: sq, theme: 'Market Outlook' });
  }

  // Real estate queries
  if (propertyTypes.length > 0) {
    queries.push({ query: 'real estate market housing prices', theme: 'Real Estate' });
    if (propertyTypes.some((pt) => pt.toLowerCase() === 'rental')) {
      queries.push({ query: 'rental property investment landlord news', theme: 'Rental Market' });
    }
  }

  // Liability queries
  if (liabilityCategories.length > 0) {
    const catsLower = liabilityCategories.filter(Boolean).map((c) => c.toLowerCase());
    if (catsLower.some((c) => c.includes('student'))) queries.push({ query: 'student loan news policy', theme: 'Student Loans' });
    if (catsLower.some((c) => c.includes('credit'))) queries.push({ query: 'credit card interest rates personal finance', theme: 'Credit Cards' });
    if (catsLower.some((c) => c.includes('auto'))) queries.push({ query: 'auto loan rates car financing', theme: 'Auto Loans' });
  }

  // Fallback
  if (queries.length === 0) {
    queries.push({ query: 'personal finance investing tips', theme: 'Personal Finance' });
  }

  return queries.slice(0, 8);
}

export async function fetchRelevantNews(
  tickers: string[] = [],
  propertyTypes: string[] = [],
  liabilityCategories: string[] = [],
  accountTypes: string[] = [],
  maxArticles: number = 8,
): Promise<NewsArticle[]> {
  // Build cache key
  const cacheInput = `${tickers}|${propertyTypes}|${liabilityCategories}|${accountTypes}`;
  const cacheKey = crypto.createHash('md5').update(cacheInput).digest('hex');

  // Check cache
  const now = Date.now();
  const cached = _newsCache.get(cacheKey);
  if (cached && now - cached.fetchedAt < NEWS_CACHE_TTL) {
    return cached.articles;
  }

  const queries = buildSearchQueries(tickers, propertyTypes, liabilityCategories, accountTypes);

  const allArticles: NewsArticle[] = [];
  const seenTitles = new Set<string>();

  for (const qInfo of queries) {
    const articles = await fetchRss(qInfo.query, 3);
    for (const article of articles) {
      const titleKey = article.title.toLowerCase().replace(/[^a-z0-9]/g, '').slice(0, 50);
      if (!seenTitles.has(titleKey)) {
        seenTitles.add(titleKey);
        article.theme = qInfo.theme;
        allArticles.push(article);
      }
    }
  }

  const result = allArticles.slice(0, maxArticles);

  // Cache result
  _newsCache.set(cacheKey, { articles: result, fetchedAt: now });

  // Prune old cache entries
  if (_newsCache.size > 50) {
    for (const [k, v] of _newsCache.entries()) {
      if (now - v.fetchedAt > NEWS_CACHE_TTL) {
        _newsCache.delete(k);
      }
    }
  }

  return result;
}
